File: 0863-all-nodes-distance-k-in-binary-tree/0863-all-nodes-distance-k-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode getTarget(TreeNode node,TreeNode target,HashMap<Integer,TreeNode> bm) {
        //base case
        if(node==null) return null;
        if(node.val==target.val) return node;
        TreeNode left = getTarget(node.left,target,bm);
        if(left!=null) {
            bm.put(node.left.val,node);
            return left;
        }
        TreeNode right = getTarget(node.right,target,bm);
        if(right!=null) {
            bm.put(node.right.val,node);
            return right;
        }
        
        return null;
    }
    public List<Integer> distanceK(TreeNode root, TreeNode target, int k) {
        // backmap,queue,set
        HashSet<Integer> vis= new HashSet<>();
        HashMap<Integer,TreeNode> bm = new HashMap<>();
        LinkedList<TreeNode> que = new LinkedList<>();
        //get targetNode
        TreeNode tar = getTarget(root,target,bm);
        que.addLast(tar);
        //BFS
        while(k-->0) {
            int size = que.size();
            while(size-->0) {
                TreeNode rn = que.removeFirst();
                
                //add left and right children
                if(rn.left != null && vis.add(rn.left.val)) {
                    que.addLast(rn.left);
                }
                if(rn.right!=null && vis.add(rn.right.val)) {
                    que.addLast(rn.right);
                }
                //explore parent paths if exist
                if(bm.containsKey(rn.val)) {
                    que.addLast(bm.get(rn.val));
                    vis.add(rn.val);
                }
            }
        }
        //add all the set nodes inside our que from bfs
        List<Integer> ans = new ArrayList<>();
        for(TreeNode node : que) {
            ans.add(node.val);
        }
        return ans;
    }
}

File: 0863-all-nodes-distance-k-in-binary-tree/0863-all-nodes-distance-k-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* dfs(TreeNode* node,TreeNode* target,unordered_map<int,TreeNode*>& bm) {
        if(!node) return NULL;
        if(node->val == target->val) return node;
        TreeNode* left = dfs(node->left,target,bm);
        if(left) {
            bm[node->left->val] = node;
            return left;
        }
        TreeNode* right = dfs(node->right,target,bm);
        if(right) {
            bm[node->right->val] = node;
            return right;
        }
        return NULL;
    }
    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
        //set , map , que
        set<int> vis;
        unordered_map<int,TreeNode*> bm;
        queue<TreeNode*> que;
        //get target
        TreeNode* tar = dfs(root,target,bm);
        cout<<tar->val;
        //BFS
        que.push(tar);
        while(k-->0) {
            int size = que.size();
            while(size-->0) {
                TreeNode* rn = que.front(); que.pop();
                //left and right bfs
                if(rn->left && vis.find(rn->left->val)==vis.end()) {
                    que.push(rn->left);
                    vis.insert(rn->left->val);
                }
                
                if(rn->right && vis.find(rn->right->val)==vis.end()) {
                    que.push(rn->right);
                    vis.insert(rn->right->val);
                }
                //EXPLORE PARENT
                if(bm.find(rn->val)!=bm.end()) {
                    que.push(bm[rn->val]);
                    vis.insert(rn->val);
                }
            }
        }
        vector<int> ans;
        while(que.size()>0) {
            ans.push_back(que.front()->val);
            que.pop();
        }
        return ans;
    }
};